src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,537 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 intelliGo Corporation. All rights reserved.
+//
+// Author: Fu-Yun TSUO <fy.tsuo@intelli-go.com>
+
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/igo_nr.h>
+#include <user/trace.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/igo_nr/igo_nr_comp.h>
+#include <sof/trace/trace.h>
+#include <sof/ut.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_igo_nr;
+
+/* 696ae2bc-2877-11eb-adc1-0242ac120002 */
+DECLARE_SOF_RT_UUID("igo-nr", igo_nr_uuid,  0x696ae2bc, 0x2877, 0x11eb, 0xad, 0xc1,
+		    0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(igo_nr_tr, SOF_UUID(igo_nr_uuid), LOG_LEVEL_INFO);
+
+static void igo_nr_capture_s16(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int16_t *x;
+	int16_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s16(source, idx_in);
+		cd->in[cd->in_wpt] = (*x);
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]);
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]);
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static void igo_nr_capture_s32(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int32_t *x;
+	int32_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = (*x) >> 16;
```

Comment:
Perhaps use round here?

Reply:
we always take the MSB 16bit regardless of input frame format being S32_LE, S24_LE, or S16_LE.

Reply:
The shift without round when going from larger world length to smaller creates a tiny DC offset to signal but it's up to you to decide.

Reply:
Ah thanks for the reminding. We are aware of quantization noise induced by bit shift. It's totally fine for our algorithm.

Reply:
@singalsu my apologies. the tiny DC offset in your statement is due to biased rounding, not quant noise. We will have a short discussion about it internally.

Reply:
@singalsu we decide to add a unbias shift prior to copying to data buffer. thanks!

Reply:
There's e.g. macro Q_SHIFT_RND() in sof/audio/format.h that does it.

Reply:
thanks. replaced with Q_SHIFT_RND().